docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,775 @@
+
+- `format_id`
+  - Each plugin needs a unique identifier to distinguish itself.
+- `on_register` and `on_unregister`
+  - A plugin might require some setup or teardown when it is registered or
+    unregistered, so these functions are required. Furthermore, a plugin
+    might require configuration, which is why there is a `config_data`
+    parameter. The configuration data can be plugin specific, so no format is
+    specified in this proposal.
+- `get_evidence`, `free_evidence`, and `free_endorsements`
+  - Producing evidence and endorsements is necessary for attestation.
+  - `flags` field to determine local vs. remote attestation.
+  - There is a `custom_claims` parameter because many attestation protocols
+    require the enclave to sign some data from a relying party. For example,
+    many protocols follow the "challenge response" architecture, which requires
+    the enclave to sign a nonce from the relying party.
+  - There is an `opt_params` field because some plugins might require plugin
+    specific input. For example, the SGX local attestation needs the
+    other enclave's target info struct.
+  - There is an `endorsements` parameter to return the endorsements that are
+    coupled with the evidence to ensure that the evidence and endorsements are
+    in sync.
+- `verify_evidence`
+  - Verifying evidence and endorsements is essential for attestation.
+  - The `claims` field contains key-value pairs that can be verified by the
+    caller. This will have the similar contents as the `oe_identity_t` field
+    in the `oe_report_t` struct returned by `oe_verify_report` and any custom
+    claims that were passed to the `get_evidence` function.
+  - The `input_validation_time` is needed to validate the evidence against a
+    specific time.
+
+###  Known Open Enclave Claims
+
+- Each plugin's `verify_evidence` function must, at minimum, return the
+  following claims (mapped from the `oe_identity_t`):
+  
+| Claim Name       | Claim Value Type   | Description                                                          |
+|:-----------------|:-------------------|:---------------------------------------------------------------------|
+| id_version       | uint32_t           | Claims version. Must be 0                                            |
+| security_version | uint32_t           | Security version of the enclave. (ISVN for SGX).                     |
+| attributes       | uint64_t           | Attributes flags for the evidence: <br/> `OE_REPORT_ATTRIBUTES_DEBUG`: The evidence is for a debug enclave.<br/> `OE_REPORT_ATTRIBUTES_REMOTE`: The evidence can be used for remote attestation.   |
+| unique_id        | uint8_t[32]        | The unique ID for the enclave (MRENCLAVE for SGX).                   |
+| signer_id        | uint8_t[32]        | The signer ID for the enclave (MRSIGNER for SGX).                    |
+| product_id       | uint8_t[32]        | The product ID for the enclave (ISVPRODID for SGX).                  |
```

Comment:
For OPTEE, we also keep a note of trusted component public key [oe_identity_t.public_key](https://github.com/openenclave/openenclave/blob/feature.new_platforms/include/openenclave/bits/report.h). This change is not yet in `master` afaik, but wanted to start a discussion if there was something similar that would be interesting to map from SGX.

Reply:
We have `oe_get_public_key` that works for SGX, so I guess that could work for both. Is the public key distinct for each platform? In other words, if the enclave gets moved to a different machine, does the enclave's public key change? If so, then it's equivalent to SGX, since we derive the asymmetric keys from the sealing keys which are different for each platform.